### ISSUE TITLE
Date Math incorrectly returns documents outside the range

### ISSUE THREAD

**Issue by cawoodm** [elastic/elasticsearch#40581]
ELK 6.6.1 Windows

When querying docs for yesterday(`>= now-1d/d AND < now/d`), docs for today are returned.

```
GET /myindex/_search?sort=timestamp:desc
{
  "from" : 0, "size" : 100,
  "query": {
    "bool": {
      "must": [
        {
          "range": {
            "timestamp": {
                "gte": "now-1d/d",
                "lt": "now/d"
            }
          }
        }
      ]
    }
  }
}
```

A document of timestamp today 00:23:57.084 (CET) is returned:
```
   "_id": "Z4d3wWkBV0vWSMeR5iew",
   "_source": {
      "timestamp": "2019-03-27T23:23:57.084Z",
   ...
```

![image](https://user-images.githubusercontent.com/498834/55149723-fb79bd00-514a-11e9-8e93-b7842dbcec17.png)

**Comment by javanna** [comment]
This looks like a timezone issue, times are stored as UTC within Elasticsearch, and this document has timestamp from 23:23:57 which does match your range query. The UI seems to apply a timezone `+1` to it when visualizing it though. Note that you can provide a [timezone](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#_time_zone_in_range_queries) as part of the range query if you want to leave your query as is, otherwise you would have to convert your range to UTC manually. Does this make sense to you?

**Comment by cawoodm** [comment]
I will try this (timezone) on monday and get back to you. For now I am wondering what timezone ES is defaulting to? Our server is CET, our client is CET, we never configured anything so I would expect a sensible default and not, say EST or GMT.

**Comment by cbuescher** [comment]
> Our server is CET, our client is CET, we never configured anything so I would expect a sensible default and not, say EST or GMT.

That is UTC, like @javanna already mentiond. More info in the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html#_timezone). Also I just checked that `timezone`is supported in date math expressions with rounding in `range` queries like in the example you mentioned.

**Comment by cawoodm** [comment]
OK, I can confirm that works.
```
"timestamp": {
    "gte": "now-1d/d",
    "lt": "now-0d/d",
    "time_zone": "+01:00"
}
```

**Comment by cawoodm** [comment]
Little help? Just trying to do a query with local dates (CET). I'm getting ` "failed to parse date field [2019-03-31 00:00:00] with format [date_time||epoch_millis]"`. I've tried `"gte": "2019-03-31 00:00:00"`, `"gte": "2019-03-31 00:00:00.000"`, `"gte": "2019-03-31 00:00:00.0000000000"` and `T` as a separator to no avail..

**Comment by spinscale** [comment]
please see the documentation for date formats in order to find out how a specific date format is structured: https://www.elastic.co/guide/en/elasticsearch/reference/6.6/mapping-date-format.html#built-in-date-formats

**Comment by cawoodm** [comment]
I did that, that's why I'm confused:

> date_time or strict_date_time
A formatter that combines a full date and time, separated by a T: yyyy-MM-dd'T'HH:mm:ss.SSSZZ.

IN: `2019-03-30T00:00:00.00000`
OUT: ERROR: `failed to parse date field [2019-03-30T00:00:00.000] with format [date_time||epoch_millis]`

**Comment by spinscale** [comment]
your example in the [above comment](https://github.com/elastic/elasticsearch/issues/40581#issuecomment-478461742) showed a completely different date format, thus my confusion. Can you please create a fully reproducible example with the new values your are using, including index creation/mapping and the full query?

Thanks a lot!

**Comment by cawoodm** [comment]
TLDR: I guess we just have to add the timezone to numeric dates - it doesn't get it from the `time_zone`.

The mapping is "date_time" and acc. the docs the only format it accepts is `yyyy-MM-ddThh:mm:ss.SSSZ` but I want to specify a local time so I dropped the `Z` and added `time_zone` to no avail.

The following works:

```json
{
    "query": {
        "bool": {
            "filter": [
                {
                    "range": {
                        "timestamp": {
                            "gte": "2019-03-30T00:00:00.000+01:00",
                            "lt": "2019-04-01T00:00:00.000+01:00",
                            "time_zone": "+01:00"
                        }
                    }
                }
            ]
        }
    }
}
```

But not this:
```json
{
    "query": {
        "bool": {
            "filter": [
                {
                    "range": {
                        "timestamp": {
                            "gte": "2019-03-30T00:00:00.000",
                            "lt": "2019-04-01T00:00:00.000",
                            "time_zone": "+01:00"
                        }
                    }
                }
            ]
        }
    }
}
```
